nibabel/filebasedimages.py

```diff
@@ -538,7 +540,30 @@ class SerializableImage(FileBasedImage):
     """
 
     @classmethod
-    def from_bytes(klass, bytestring):
+    def _filemap_from_iobase(klass, ioobject: io.IOBase):
+        """For single-file image types, make a file map with the correct key"""
+        if len(klass.files_types) > 1:
+            raise NotImplementedError(
+                "(de)serialization is undefined for multi-file images"
+            )
+        return klass.make_file_map({klass.files_types[0][0]: ioobject})
+
+    @classmethod
+    def _from_iobase(klass, ioobject: io.IOBase):
+        """Load image from readable IO stream
+
+        Convert to BytesIO to enable seeking, if input stream is not seekable
+        """
+        if not ioobject.seekable():
```

Comment:
Do we need a specific error if the stream is not readable?

Reply:
This is what we'll currently get:

```Python
In [1]: f = open('deleteme', 'w')

In [2]: f.read()
---------------------------------------------------------------------------
UnsupportedOperation                      Traceback (most recent call last)
Input In [2], in <cell line: 1>()
----> 1 f.read()

UnsupportedOperation: not readable
```

Any thoughts on a better exception than `io.UnsupportedOperation`?

Reply:
But - I guess we'll get that in due course, either here, if the object is not `seekable`, or later when we try and read from the image?   Do you think there's any advantage here for checking if the stream is readable and raising here, as opposed to later?

Reply:
Well, let's just look at what we get:

```Python
In [7]: nb.Nifti1Image.from_stream(f)
---------------------------------------------------------------------------
UnsupportedOperation                      Traceback (most recent call last)
<ipython-input-7-174c60215a44> in <module>
----> 1 nb.Nifti1Image.from_stream(f)

~/Projects/nipy/nibabel-mnt/nibabel/filebasedimages.py in from_stream(klass, io_obj)
    557         if not io_obj.seekable():
    558             io_obj = io.BytesIO(io_obj.read())
--> 559         return klass.from_file_map(klass._filemap_from_iobase(io_obj))
    560 
    561     def to_stream(self, io_obj: io.IOBase, **kwargs):

~/Projects/nipy/nibabel-mnt/nibabel/analyze.py in from_file_map(klass, file_map, mmap, keep_file_open)
    967         hdr_fh, img_fh = klass._get_fileholders(file_map)
    968         with hdr_fh.get_prepare_fileobj(mode='rb') as hdrf:
--> 969             header = klass.header_class.from_fileobj(hdrf)
    970         hdr_copy = header.copy()
    971         imgf = img_fh.fileobj

~/Projects/nipy/nibabel-mnt/nibabel/nifti1.py in from_fileobj(klass, fileobj, endianness, check)
    684     @classmethod
    685     def from_fileobj(klass, fileobj, endianness=None, check=True):
--> 686         raw_str = fileobj.read(klass.template_dtype.itemsize)
    687         hdr = klass(raw_str, endianness, check)
    688         # Read next 4 bytes to see if we have extensions.  The nifti standard

~/Projects/nipy/nibabel-mnt/nibabel/openers.py in read(self, *args, **kwargs)
    191 
    192     def read(self, *args, **kwargs):
--> 193         return self.fobj.read(*args, **kwargs)
    194 
    195     def readinto(self, *args, **kwargs):

UnsupportedOperation: not readable
```

It's three layers deeper in the call stack, but the traceback is pretty readable. My preference is to only preempt errors if they're going to be difficult to interpret when the downstream user gets them, so I'm inclined to keep this one as it is. Unless you'd rather hide the `ImageOpener()` machinery?

Reply:
No - that's fine - I'm happy if you're happy.